ce/api/streamflow/watershed.py

```diff
@@ -0,0 +1,421 @@
+        self.units = units
+        self.lon_step = longitudes[1] - longitudes[0]
+        self.lat_step = latitudes[1] - latitudes[0]
+
+    @staticmethod
+    def from_nc_dataset(dataset, variable_name):
+        """Factory method. Extracts relevant data from a netcdf file (`Dataset`)
+        with standard contents and returns it as a `DataGrid`."""
+        return DataGrid(
+            dataset.variables['lon'],
+            dataset.variables['lat'],
+            dataset.variables[variable_name],
+            dataset.variables[variable_name].units,
+        )
+
+    def is_compatible(self, other):
+        """Return a boolean indicating whether this `DataGrid` and another are
+        compatible. Compatible means that their lon and lat grids have the same
+        step size."""
+        return math.isclose(self.lon_step, other.lon_step) and \
+               math.isclose(self.lat_step, other.lat_step)
+
+    def check_valid_index(self, index):
+        if not is_valid_index(index, self.values.shape):
+            raise DataGridIndexError(index, self.values.shape)
+
+    def lonlat_to_xy(self, lonlat):
+        """Returns the (x, y) data index for a given lon-lat coordinate,
+        switching the order of the coordinates. Checks that the index is
+        valid for the grid; we must at minimum exclude negative index values,
+        which are valid but wrong in this application."""
+        x = int(round((lonlat[1] - self.latitudes[0]) / self.lat_step))
+        y = int(round((lonlat[0] - self.longitudes[0]) / self.lon_step))
+        self.check_valid_index((x, y))
+        return x, y
+
+    def xy_to_lonlat(self, xy):
+        """Returns the lon-lat coordinate for a given xy data index,
+        switching the order of the coordinates."""
+        self.check_valid_index(xy)
+        return self.longitudes[xy[1]], self.latitudes[xy[0]]
+
+    def get_values_at_lonlats(self, lonlats):
+        """Map an iterable of lonlats to a list of values at those lonlats"""
+        return [
+            float(self.values[self.lonlat_to_xy(lonlat)]) for lonlat in lonlats
+        ]
+
+
+def watershed(sesh, station, ensemble_name):
+    """Return information describing the watershed that drains to a specified
+    point.
+
+    :param sesh: (sqlalchemy.orm.session.Session) A database Session object
+    :param station: (string) Location of drainage point, WKT POINT format
+    :param ensemble_name: (string) Name of the ensemble containing data files backing
+        providing data for this request.
+    :return: (dict) representation for JSON response object. See function
+        `worker` for details.
+
+    This function is primarily responsible for finding the relevant data files
+    and converting their contents to `DataGrid` objects for consumption by
+    `worker`, which as its name suggests, does most of the work.
+    """
+    station_lonlat = WKT_point_to_lonlat(station)
+
+    with get_time_invariant_variable_dataset(
```

Comment:
I think that you can specify multiple context managers in a single with statement to avoid the massive indent. I.e.

```
with foo() as bar, blah() as stuff:
    do_something_with(bar, stuff)
```

Reply:
Oh, man, I was wishing for that.